unit/util/expr_initializer.cpp

```diff
@@ -0,0 +1,274 @@
+// Author: Diffblue Ltd.
+
+#include <util/arith_tools.h>
+#include <util/bitvector_types.h>
+#include <util/c_types.h>
+#include <util/expr_initializer.h>
+#include <util/namespace.h>
+#include <util/std_code.h>
+#include <util/symbol_table.h>
+
+#include <testing-utils/use_catch.h>
+
+/// Helper struct to hold useful test components.
+struct expr_initializer_test_environmentt
+{
+  symbol_tablet symbol_table;
+  namespacet ns{symbol_table};
+  source_locationt loc{};
+
+  static expr_initializer_test_environmentt make()
+  {
+    return {};
+  }
+
+private:
+  expr_initializer_test_environmentt() = default;
+};
+
+/// Helper function to create and populate a c_enum_typet type.
+static c_enum_typet make_c_enum_type(
+  const unsignedbv_typet &underlying_type,
+  unsigned int value_count)
+{
+  c_enum_typet enum_type{underlying_type};
+
+  auto &members = enum_type.members();
+  members.reserve(value_count);
+
+  for(unsigned int i = 0; i < value_count; ++i)
+  {
+    c_enum_typet::c_enum_membert member;
+    member.set_identifier("V" + std::to_string(i));
+    member.set_base_name("V" + std::to_string(i));
+    member.set_value(integer2bvrep(i, underlying_type.get_width()));
+    members.push_back(member);
+  }
+  return enum_type;
+}
+
+/// Create a type symbol, a type_tag and a symbol for the tag and populate with
+/// them the symbol_table in the environment.
+static symbolt create_tag_populate_env(
+  const typet &type,
+  expr_initializer_test_environmentt &env)
+{
+  const type_symbolt symbol{"my_type_symbol", type, ID_C};
```

Comment:
:pick: If this function were to be called multiple times in the same test, then it would be inserting duplicate symbol names into the symbol table. This seems like a potential trap when maintaining these tests in future. Consider using the existing `get_fresh_aux_symbol` from `src/util/fresh_symbol.h` in order to mitigate this risk.

Reply:
Could be, but that function requires many parameters that its usage may make the test more confusing.

Reply:
No.
The `get_fresh_aux_symbol` function will ALWAYS return a "normal" `symbolt`, and not a `type_symbolt`, so then the lookup in the namespace will fail as that requires the resolved `symbolt` to be a `type_symbol`.